zebra/zapi_msg.c

```diff
@@ -2025,6 +2136,7 @@ static void zread_hello(ZAPI_HANDLER_ARGS)
 	if (!client->synchronous) {
 		zsend_capabilities(client, zvrf);
 		zebra_vrf_update_all(client);
+		zebra_srv6_locator_update_all(client);
```

Comment:
Why does the srv6 information need to be sent immediately (and without a client's asking) - why can't interested clients ask for the current info as they do for other things?

Reply:
In bgpd's SRv6 VPNv4 configration, specify the locator in the BGP instance that advertises the route of VPNv4.

At this time, it was designed to synchronize what kind of locators that zebra holds in advance, cache it on the BGPd side, and refer to it directly.

This is required when loading the startup config of BGPd.

The CLI will be as follows, which will be provided by my next PR.

```
router bgp 1
 (snip)
 !
 segment-routing srv6
  locator LOC1
!
router bgp 1 vrf vrf1
 (snip)
!
segment-routing
 srv6
  locators
   locator LOC1
    prefix 2001:db8:a:a::/64
   !
  !
 !
!
```

Reply:
I'm still trying to follow this, I guess.

if only bgp needs some information, then ... bgp should ask for it. I don't (personally) care to have every daemon be forced to receive information that it may not (ever) use?

we have other examples where a daemon has to tolerate a configuration token or keyword for some object that may not yet exist. in your example about bgp, is it always possible to guarantee that all information from zebra will be received and processed before the bgp configuration is delivered?

Reply:
Maybe a more-general issue question is worth asking: you've placed responsibility for srv6 config info in zebra alone, so zebra has to distribute this config info to other daemons. we have other examples where a library module exposes and manages its configuration, in the context of each daemon using the library feature/api. should srv6 info be managed that way too?

Reply:
> if only bgp needs some information, then ... bgp should ask for it. I don't (personally) care to have every daemon be forced to receive information that it may not (ever) use?

In correct, "routing protocol that uses SRv6" need SRv6-locator/function info. These are bgp,isis,ospf at now. (thus it's not all routing protocols)

And current zapi-client mechanism, it's not involved something problem. the message which is not interested by zapi-client will be ignored only (by null function pointer). but some communication overhead will be in there.

Broadcasting mechanism is used to know which SRv6-locator/function are exist by routing-protocols(zclient).

But making broadcast domain smaller sounds good for me too. If you know any good example, could you show me it. (I guess, zclient flags..?)

> we have other examples where a daemon has to tolerate a configuration token or keyword for some object that may not yet exist. 

Umm, that is one of the way. and I don't have any strong intent which is better. Could you please tell me the example? Is it VRF for bgp instance?

> in your example about bgp, is it always possible to guarantee that all information from zebra will be received and processed before the bgp configuration is delivered?
Yes. bgpd establish the zapi connection before configuration loading.
So it is guaranteed. (If it's wrong, please tell me that. :bow: )

> you've placed responsibility for srv6 config info in zebra alone, so zebra has to distribute this config info to other daemons. 

Distribution is performed by bcast messaging as async notification.

> we have other examples where a library module exposes and manages its configuration, in the context of each daemon using the library feature/api. should srv6 info be managed that way too?

Yes, srv6 info should be managed as well.
Another routing protocol need to allocate new SRv6-SID from the SRv6-locator. ZEBRA_SRV6_FUNCTION_ADD can be used for that.

All SRv6-locator/function can be know from the bcast message. Routing protocol only need to send new SID allocation request.

Reply:
@mjstapp Thank you for your great comments. That's really good to learn FRRouting more. :)

Reply:
This topic (broadcasting infos) is only occurred when operator uses SRv6 feature on zebra.
So it doesn't have big impact for most cases. I want to work this topic after next PR or future work.

How do you feel that?